### PR TITLE
local_dir for distributed LocalCluster

### PR DESCRIPTION
**Pull request by earthgecko**
Fix https://github.com/blue-yonder/tsfresh/issues/371

Pass tempfile local_dir to distributed LocalCluster

And Correct shell syntax in .travis.yml as per documented in #370

### CODE REVIEW

**Comment by earthgecko** [comment]
I have merged https://github.com/blue-yonder/tsfresh/pull/370 pr #370 with this pr

**Comment by earthgecko** [comment]
Travis passes

```
[gw0] [ 88%] PASSED tests/units/utilities/test_distribution.py::LocalDaskDistributorTestCase::test_local_dask_cluster_extraction
```
https://travis-ci.org/blue-yonder/tsfresh/jobs/356867648#L1267

**Comment by MaxBenChrist** [comment]
The python 3 build fails due to a time out, any idea why?

**Comment by earthgecko** [comment]
I will have a dig and debug it and get back to you :)

**Comment by earthgecko** [comment]
Hi @MaxBenChrist 

test_local_dask_cluster_extraction on Python 3.5.2 is working, after job was ```Running for 12 min 25 sec``` test_local_dask_cluster_extraction passed at https://travis-ci.org/blue-yonder/tsfresh/jobs/357472296#L1199

```
tests/units/utilities/test_distribution.py::LocalDaskDistributorTestCase::test_local_dask_cluster_extraction 
[gw0] PASSED tests/units/utilities/test_distribution.py::LocalDaskDistributorTestCase::test_local_dask_cluster_extraction 
[gw1] FAILED tests/integrations/test_notebooks.py::NotebooksTestCase::test_basic_pipeline_example 
tests/integrations/test_notebooks.py::NotebooksTestCase::test_friedrich_coefficients 
```

test_basic_pipeline_example failed ... unrelated I think.

It works locally for me on Python 3.5.2, I have reconfirmed building an environment from scratch (but with Sphinx==1.6.4 due to #373).

So I think the timeout was just a 10m 0 second timeout limit on Travis probably, a slow test instance?

However, the test that failed was related to another timeout - https://travis-ci.org/blue-yonder/tsfresh/jobs/357472296#L1397

```TimeoutError: Cell execution timed out```

Basically taking too long for the defaults?

**Comment by earthgecko** [comment]
Basically in _netbook_run ```"--ExecutePreprocessor.timeout=600"]``` is not long enough on Travis for some tests, sometimes.

https://github.com/blue-yonder/tsfresh/blob/6db60f46e3587540b81d7b0187dc46047e40a15e/tests/integrations/test_notebooks.py#L25

So a new issue and PR :)

**Comment by earthgecko** [comment]
Looks like it is a Travis 10min timeout ....

https://travis-ci.org/blue-yonder/tsfresh/jobs/357506408#L1201

```
[gw0] PASSED tests/units/utilities/test_distribution.py::LocalDaskDistributorTestCase::test_local_dask_cluster_extraction 
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

So even though the notebook has been given 900 seconds, Travis expects output after 10 mins.

I will https://docs.travis-ci.com/user/common-build-problems/ are add some shell environment variables, to be continued ... :)

**Comment by earthgecko** [comment]
## The pros and cons (and notebooks what about them?)

So adding travis_wait ```travis_wait 20 python setup.py test``` makes a bit of a difference, however it comes at a price.

And the Python-3.5.2 build still failed on

```
[gw1] FAILED tests/integrations/test_notebooks.py::NotebooksTestCase::test_basic_pipeline_example 
```

For reference build jobs py2 - https://travis-ci.org/blue-yonder/tsfresh/jobs/357545569 and py3 - https://travis-ci.org/blue-yonder/tsfresh/jobs/357545570

### cons

1. You lose the real time output in the Travis jobs page.  Basically Travis spits out ALL the output only at the end in one big paste and for the whole job just sits on


```
The command "sed -i 's/\-n auto/\-n 2/g' setup.cfg" exited with 0.
$ if [ $TRAVIS_PULL_REQUEST == false ] && ! [ $TRAVIS_BRANCH == 'master' ]; then py.test tests/units -n2; else travis_wait 20 python setup.py test; fi
```

Then jumps to 

```
================== 227 passed, 14 warnings in 526.73 seconds ===================
/home/travis/.travis/job_stages: line 169:  3473 Terminated              travis_jigger $! $timeout $cmd
The command "if [ $TRAVIS_PULL_REQUEST == false ] && ! [ $TRAVIS_BRANCH == 'master' ]; then py.test tests/units -n2; else travis_wait 20 python setup.py test; fi" exited with 0.
cache.2
store build cache
0.01s
33.66schange detected (content changed, file is created, or file is deleted):
/home/travis/.cache/pip/http/0/1/5/1/0/01510501120df000abad3969c9311e384214fa17d7027ef95bf91707
/home/travis/.cache/pip/http/0/4/6/f/2/046f263a04464ad173f2b55b53089d6611b281fd92901c6030b4cd93
/home/travis/.cache/pip/http/0/5/1/d/d/051ddbcfada687fb27df4b43966167a67f6706da37bcbb89d2838836
/home/travis/.cache/pip/http/0/6/0/7/b/0607bfec44b45165c2faff761ccac99a98d9fd850746c7745c710bfa
/home/travis/.cache/pip/http/0/c/5/e/4/0c5e4121c33711c362362e34cc3a3697cef22b9e6f74a0440131f791
/home/travis/.cache/pip/http/0/e/7/b/9/0e7b99547a0247a24df2175f35e166cf96f0194417e99b4851dbf306
/home/travis/.cache/pip/http/1/7/b/b/4/17bb4c359c58b8944aba924db1f952f1ebf83975f86088985f70f836
/home/travis/.cache/pip/http/1/8/a/f/f/18affb199964ddd5181e75313e322994f424d8d77b78d0192a7e0d15
/home/travis/.cache/pip/http/1/8/d/e/0/18de0670993434c1d60d77f59723689f00d0e2e81d36e0202f1f6459
/home/travis/.cache/pip/http/1/b/b/d/e/1bbde7f10876562f138fa536953ad413bfc645c4f5f79dd4f103eae5
/home/travis/.cache/pip/http/2/1/8/a/6/21
...
changes detected, packing new archive
.
.
.
.
uploading archive
after_success
1.99s$ coveralls
Submitting coverage to coveralls.io...
Coverage submitted!
Job #1339.1
https://coveralls.io/jobs/34948709
Done. Your build exited with 0.
```
2. You lose the green PASSED and red FAILED - in the single output there is NO colour, which is pretty shit as when you want to know which of the 227 tests FALIED, it is easy to find in red.

3. It seems to lose the timeout or error messages one has come to expect.

```
tests/integrations/test_notebooks.py::NotebooksTestCase::test_robot_failure_example 
The command "if [ $TRAVIS_PULL_REQUEST == false ] && ! [ $TRAVIS_BRANCH == 'master' ]; then py.test tests/units -n2; else travis_wait 20 python setup.py test; fi" exited with 137.
cache.2
store build cache
0.13s
158.74schange detected (content changed, file is created, or file is deleted):
...
...
uploading archive
Done. Your build exited with 1.
```

### pros

It would make tests pass, but I think ```ExecutePreprocessor.timeout``` would have to be increased more than 900 as on Python-3.5.2 NotebooksTestCase::test_robot_failure_example still fails even though ```travis_wait 20```


### Possible other options or solutions:

- Split out the notebook tests to another ```tests``` run and run 2 stages on Travis, not sure if that is possible.
- Only run the notebook tests on developer local testing

Testing using the environment variable ```TRAVIS=true``` as per https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables

Running ***with*** notebook tests takes 304.19 seconds (on my machine)
Running ***without*** notebook tests takes 42.23 seconds (on my machine)

So this is possible in tests/integrations/test_notebooks.py

```
    if os.environ['TRAVIS']:
        print('Not running notebook tests on Travis')
        return [], []

    with tempfile.NamedTemporaryFile(mode='w+t', suffix=".ipynb") as fout:
        args = ["jupyter", "nbconvert",
```

But is it desirable?

- It would make user PR builds faster.
- It would put the onus on the merger to ensure that notebook tests were run if a notebook was added or changed.

So the question is, do notebook tests need to be run on Travis?  Considering
that every tsfresh function called within a notebook probably already has a test
within the tsfresh context, can it be assumed that on Travis the notebook would
pass?  I guess there could be breaking changes in dependencies that could break
notebooks as well.  However seeing as the developers should be running tests
locally themselves all the time before committing one would hope, they will be
running the notebook tests so the notebook tests will still get run.

Not ideal at all, however it does significantly reduce Travis build times and
would remove a lot of wait time from general public PRs.

Anyway thoughts @MaxBenChrist ...

**Comment by earthgecko** [comment]
So Travis job pass if notebook testing is excluded - https://travis-ci.org/blue-yonder/tsfresh/builds/357811821

Going to test with 2.7.14 and 3.6.4 ***with*** notebook tests, 3.6.4 may be speedier.

**Comment by earthgecko** [comment]
OK Python 2.7.14 and 3.6.4 both managed to get threw all the tests :)  But will documentation?

To be continued...

**Comment by MaxBenChrist** [comment]
Sorry for the super delayed response, @earthgecko. Your changes look good!

I think the best way is to move the notebooks into its own stage.
So far we either only test tests/units or tests/. We should split that up into two stages tests/unit and tests/integration.

I am not sure if I want to ramp up the python versions for the overall testing.

**Comment by earthgecko** [comment]
Now a new issue when the tests passed in March statsmodels was at 0.8.0 now statsmodels is at 0.9.0 and is throwing an error in augmented_dickey_fuller- https://travis-ci.org/blue-yonder/tsfresh/jobs/391340091#L1468

Which is a new issue, only way to fix this now is to pin in requirements.

**Comment by earthgecko** [comment]
I can reproduce Travis errors locally nowon a fresh virtualenv, so something else in requirements,txt moved too.  Back to diffing travis pip freeze out travis.1342.1.passed.pip.freeze.txt/travis.1431.1.failed.pip.freeze.txt

```
============================================================================================================== FAILURES ===============================================================================================================
_______________________________________________________________________________________ RollingTestCase.test_warning_on_non_uniform_time_steps ________________________________________________________________________________________
[gw2] linux -- Python 3.5.2 /opt/python_virtualenv/projects/tsfresh-pr372-py352/tsfresh/../bin/python3.5

self = <tests.units.utilities.test_dataframe_functions.RollingTestCase testMethod=test_warning_on_non_uniform_time_steps>

    def test_warning_on_non_uniform_time_steps(self):
        with warnings.catch_warnings(record=True) as w:
            first_class = pd.DataFrame({"a": [1, 2, 3, 4], "b": [5, 6, 7, 8], "time": [1, 2, 4, 5]})
            second_class = pd.DataFrame({"a": [10, 11], "b": [12, 13], "time": range(20, 22)})

            first_class["id"] = 1
            second_class["id"] = 2

            df_full = pd.concat([first_class, second_class], ignore_index=True)

            dataframe_functions.roll_time_series(df_full, column_id="id", column_sort="time",
                                                 column_kind=None, rolling_direction=1)

>           self.assertEqual(len(w), 1)
E           AssertionError: 3 != 1

/opt/python_virtualenv/projects/tsfresh-pr372-py352/tsfresh/tests/units/utilities/test_dataframe_functions.py:408: AssertionError

```

**Comment by MaxBenChrist** [comment]
this gets confusing, can we split the pr into smaller prs? I would just merge the dask tempfile, then we look over the notebook timeout issue in another pr, the statsmodel issue should also get its own pr

**Comment by MaxBenChrist** [comment]
and the issues with the documentation build should also get its own pr I guess

**Comment by earthgecko** [comment]
Hi @MaxBenChrist I thought I might get to there but did not as it would just get me back to the same place.  There are 2 changes that were required to make the tests pass, I saw you handled the dask temp dir issue, I think.  Further I got stopped at the new failures.

That said, for now the easiest way to stop breaking forward is to fix all requirements to the last known working versions (travis review required).  That way at least new users that are installing tsfresh will not be reporting somewhat expected breakages and then release v0.1.2 with only that change.  In my opinion that will hopefully solve all these new issues.

Then starting a process of one by one pip upgrades and fixing as we go, this package drift is not going to stop unless there is a somewhat concerted effort to try and get tsfresh up to date with the moving deps.

At this point is, this is going to be quite a bit of testing and bug hunting/squashing, enough time to not be trivial I would imagine.  I am going to try and get started on it, but if my prs that test and fail and are debugged and explained as best as I can are not merged, then it is not that easy to move any of these issues forward.  This is not a rant, I just honestly did not get back to splitting them all up into multiple prs and it did not really make sense to me months later and there were further errors and I got busy.

**Comment by earthgecko** [comment]
Individual PRs - as much as possible:

- increase-timeouts-notebook-tests #409 (also has commented out code to disable notebook tests on Travis if desired)
- Correct syntax in .travis.yml #411

@MaxBenChrist you have handled the dask temp dir, so those 2 prs cover the other aspect of this PR which I am now going to close and try and move forward with the deps and related bugs.
